### PR TITLE
GH-103857: Deprecate utcnow and utcfromtimestamp

### PR DESCRIPTION
**Pull request by pganssle**
Per the reasoning in #103857.

To do:

- [x] Add tests
- [x] Clean up internal uses of `utcnow` and `utcfromtimestamp`

* Issue: gh-103857

### CODE REVIEW

File: Lib/datetime.py [hunk 1]
```diff
@@ -1801,6 +1801,14 @@ def fromtimestamp(cls, timestamp, tz=None):
     @classmethod
     def utcfromtimestamp(cls, t):
         """Construct a naive UTC datetime from a POSIX timestamp."""
+        import warnings
+        warnings.warn("datetime.utcfromtimestamp() is deprecated and scheduled "
+                      "for removal in a future version. It is instead "
+                      "recommended that users use timezone-aware objects to "
+                      "represent datetimes in UTC. To get such an object from "
+                      "a timestamp, use datetime.fromtimestamp(t, datetime.UTC)",
```
Comment (erlend-aasland) [inline Lib/datetime.py]:
```suggestion
                      "a timestamp, use datetime.fromtimestamp(t, datetime.UTC).",
```

File: Lib/datetime.py [hunk 2]
```diff
@@ -1812,8 +1820,16 @@ def now(cls, tz=None):
     @classmethod
     def utcnow(cls):
         "Construct a UTC datetime from time.time()."
+        import warnings
+        warnings.warn("datetime.utcnow() is deprecated and scheduled for "
+                      "removal in a future version. It is instead recommended "
+                      "that users use timezone-aware objects to represent "
+                      "datetimes in UTC. To get such an object representing "
+                      "the current time, use datetime.now(datetime.UTC)",
```
Comment (erlend-aasland) [inline Lib/datetime.py]:
```suggestion
                      "the current time, use datetime.now(datetime.UTC).",
```

File: Modules/_datetimemodule.c [hunk 3]
```diff
@@ -5144,6 +5144,14 @@ datetime_datetime_now_impl(PyTypeObject *type, PyObject *tz)
 static PyObject *
 datetime_utcnow(PyObject *cls, PyObject *dummy)
 {
+    PyErr_WarnEx(
+        PyExc_DeprecationWarning,
+        "datetime.utcnow() is deprecated and scheduled for removal in a future "
+        "version. It is instead recommended that users use timezone-aware "
+        "objects to represent datetimes in UTC. To get such an object "
+        "representing the current time, use datetime.now(datetime.UTC)",
```
Comment (erlend-aasland) [inline Modules/_datetimemodule.c]:
```suggestion
        "representing the current time, use datetime.now(datetime.UTC).",
```

File: Modules/_datetimemodule.c [hunk 4]
```diff
@@ -5180,6 +5188,14 @@ datetime_fromtimestamp(PyObject *cls, PyObject *args, PyObject *kw)
 static PyObject *
 datetime_utcfromtimestamp(PyObject *cls, PyObject *args)
 {
+    PyErr_WarnEx(
+        PyExc_DeprecationWarning,
+        "datetime.utcfromtimestamp() is deprecated and scheduled for removal "
+        "in a future version. It is instead recommended that users use "
+        "timezone-aware objects to represent datetimes in UTC. To get such an "
+        "object from a timestamp, use datetime.fromtimestamp(t, datetime.UTC)",
```
Comment (erlend-aasland) [inline Modules/_datetimemodule.c]:
```suggestion
        "object from a timestamp, use datetime.fromtimestamp(t, datetime.UTC).",
```

**Review by erlend-aasland — APPROVED** [review]
LGTM (fixing the remaining test failures seems straight-forward.)

File: Lib/datetime.py (hunk 1 above)
Comment (pganssle) [reply Lib/datetime.py]: Done. Thanks!

File: Lib/datetime.py [hunk 5]
```diff
@@ -1801,6 +1801,14 @@ def fromtimestamp(cls, timestamp, tz=None):
...
     def utcfromtimestamp(cls, t):
         """Construct a naive UTC datetime from a POSIX timestamp."""
+        import warnings
+        warnings.warn("datetime.utcfromtimestamp() is deprecated and scheduled "
+                      "for removal in a future version. It is instead "
+                      "recommended that users use timezone-aware objects to "
+                      "represent datetimes in UTC. To get such an object from "
+                      "a timestamp, use datetime.fromtimestamp(t, datetime.UTC).",
+                      DeprecationWarning,
+                      stacklevel=2)
```
Comment (ambv) [inline Lib/datetime.py]: I find this a little wordy for a warning, and tentative. The deprecated function will be removed in the future so we are not recommending an alternative, we are *telling* people to use an alternative. How about this:

```suggestion
        warnings.warn("datetime.utcfromtimestamp() is deprecated and will be "
                      "removed in a future version. Use timezone-aware objects to "
                      "represent datetimes in UTC: "
                      "datetime.fromtimestamp(t, datetime.UTC)",
                      DeprecationWarning,
                      stacklevel=2)
```

File: Lib/datetime.py [hunk 6]
```diff
@@ -1812,8 +1820,16 @@ def now(cls, tz=None):
...
     def utcnow(cls):
         "Construct a UTC datetime from time.time()."
+        import warnings
+        warnings.warn("datetime.utcnow() is deprecated and scheduled for "
+                      "removal in a future version. It is instead recommended "
+                      "that users use timezone-aware objects to represent "
+                      "datetimes in UTC. To get such an object representing "
+                      "the current time, use datetime.now(datetime.UTC).",
+                      DeprecationWarning,
+                      stacklevel=2)
```
Comment (ambv) [inline Lib/datetime.py]: Ditto.

```suggestion
        warnings.warn("datetime.utcnow() is deprecated and will be removed "
                      "in a future version. Use timezone-aware objects to "
                      "represent datetimes in UTC: datetime.now(datetime.UTC)",
                      DeprecationWarning,
                      stacklevel=2)
```

File: Modules/_datetimemodule.c [hunk 7]
```diff
@@ -5144,6 +5144,14 @@ datetime_datetime_now_impl(PyTypeObject *type, PyObject *tz)
...
 datetime_utcnow(PyObject *cls, PyObject *dummy)
 {
+    PyErr_WarnEx(
+        PyExc_DeprecationWarning,
+        "datetime.utcnow() is deprecated and scheduled for removal in a future "
+        "version. It is instead recommended that users use timezone-aware "
+        "objects to represent datetimes in UTC. To get such an object "
+        "representing the current time, use datetime.now(datetime.UTC).",
+        2
+    );
```
Comment (ambv) [inline Modules/_datetimemodule.c]: Ditto here about replacing the warning with the shorter version.

File: Modules/_datetimemodule.c [hunk 8]
```diff
@@ -5180,6 +5188,14 @@ datetime_fromtimestamp(PyObject *cls, PyObject *args, PyObject *kw)
...
 datetime_utcfromtimestamp(PyObject *cls, PyObject *args)
 {
+    PyErr_WarnEx(
+        PyExc_DeprecationWarning,
+        "datetime.utcfromtimestamp() is deprecated and scheduled for removal "
+        "in a future version. It is instead recommended that users use "
+        "timezone-aware objects to represent datetimes in UTC. To get such an "
+        "object from a timestamp, use datetime.fromtimestamp(t, datetime.UTC).",
+        2
+    );
```
Comment (ambv) [inline Modules/_datetimemodule.c]: Ditto.

**Review by ambv — COMMENTED** [review]
I like this, my nit is about the wordiness of the warnings.

Outside of that, there's the question of mixing tz-aware and tz-naive datetimes when it comes to comparisons and arithmetic operations. It might be somewhat disruptive for users to replace one with the other to discover TypeErrors in their code due to mixing naive and tz-aware datetimes.

File: Lib/datetime.py (hunk 5 above)
Comment (pganssle) [reply Lib/datetime.py]: I did think it was a bit wordy when I wrote it. My only reservation here is that users still *can* use naïve objects as they want, they just need to do `datetime.now(UTC).replace(tzinfo=None)`.

I suppose this subtle distinction is probably going to be lost on most people, so I'll just go with your wording.

**Comment by pganssle** [comment]
@ambv (or others), what do you think of this wording?

> datetime.utcnow() is deprecated and scheduled for removal in a future version.
> Instead, Use timezone-aware objects to represent datetimes in UTC:
> datetime.now(datetime.UTC). **This is not a drop-in replacement.**

Alternatively, **This will not be a backwards-compatible change.** or something to that effect?

File: Lib/datetime.py (hunk 5 above)
Comment (erlend-aasland) [reply Lib/datetime.py]: I agree that Łukasz's suggestion is an improvement. I think it is ok to suggest the slightly disruptive change; we _want_ people to use tz-aware objects[^1].

[^1]: Providing a drop-in replacement does not promote using tz-aware objects.

Comment (pganssle) [reply Lib/datetime.py]: > I agree that Łukasz's suggestion is an improvement. I think it is ok to suggest the slightly disruptive change; we want people to use tz-aware objects

I mean, in either case we're suggesting the change, the question is how best to communicate to users that the thing we're suggesting is not a drop-in replacement for the existing workflow. I tend to assume that when I see a deprecation warning with a code snippet in it, the code snippet is how you achieve *the same thing* in the new version. Ideally we'd make it clear that we are saying, "The thing you are doing now is not something you should continue doing, here's a new way to do it but you need to understand the implications of making this change."

That said, I'm not convinced that it's possible to convey this succinctly, so probably best to just leave it at Łukasz's wording.

**Comment by pganssle** [comment]
OK, I'm going to merge this when CI passes, using Łukasz's wording, unless there are some objections.

**Review by ankit-gautam23 — COMMENTED** [review]
LGTM
